queue/queue.mbt

```diff
@@ -0,0 +1,333 @@
+/// queue.clear()
+/// ```
+pub fn clear[T](self : Queue[T]) -> Unit {
+  self.length = 0
+  self.first = Nil
+  self.last = Nil
+}
+
+test "clear" {
+  let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+  @assertion.assert_eq(queue.length, 4)?
+  queue.clear()
+  @assertion.assert_eq(queue.length, 0)?
+}
+
+pub fn length[T](self : Queue[T]) -> Int {
+  self.length
+}
+
+/// Checks if the queue is empty.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::new()
+/// let is_empty = queue.is_empty()
+/// ```
+pub fn is_empty[T](self : Queue[T]) -> Bool {
+  self.length == 0
+}
+
+test "is_empty" {
+  let queue : Queue[Int] = new()
+  @assertion.assert_true(queue.is_empty())?
+  queue.push(1)
+  queue.push(2)
+  queue.push(3)
+  queue.push(4)
+  @assertion.assert_false(queue.is_empty())?
+}
+
+/// Adds a value to the queue.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::new()
+/// queue.add(1)
+/// ```
+pub fn push[T](self : Queue[T], x : T) -> Unit {
+  let cell : Cell[T] = Cons({ content: x, next: Nil })
+  match self.last {
+    Nil => {
+      self.length = 1
+      self.first = cell
+      self.last = cell
+    }
+    Cons(last) => {
+      last.next = cell
+      self.length = self.length + 1
+      self.last = cell
+    }
+  }
+}
+
+test "push" {
+  let queue : Queue[Int] = Queue::new()
+  queue.push(1)
+  queue.push(2)
+  queue.push(3)
+  queue.push(1)
+  @assertion.assert_eq(queue.length, 4)?
+  @assertion.assert_eq(queue.pop(), 1)?
+  @assertion.assert_eq(queue.pop(), 2)?
+  @assertion.assert_eq(queue.pop(), 3)?
+  @assertion.assert_eq(queue.pop(), 1)?
+}
+
+/// Peeks at the first value in the queue.
+/// 
+/// Panics if the queue is empty.
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+/// let first = queue.peek_exn()
+/// ```
+pub fn peek_exn[T](self : Queue[T]) -> T {
+  match self.first {
+    Nil => abort("Queue is empty")
+    Cons(first) => first.content
+  }
+}
+
+test "peek_exn" {
+  let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+  @assertion.assert_eq(queue.peek_exn(), 1)?
+  @assertion.assert_eq(queue.length, 4)?
+  @assertion.assert_eq(queue.pop(), 1)?
+  @assertion.assert_eq(queue.peek_exn(), 2)?
+  @assertion.assert_eq(queue.length, 3)?
+}
+
+/// Peeks at the first value in the queue, which returns None if the queue is empty.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+/// let first = queue.peek()
+/// ```
+pub fn peek[T](self : Queue[T]) -> Option[T] {
+  match self.first {
+    Nil => None
+    Cons(first) => Some(first.content)
+  }
+}
+
+test "peek" {
+  let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+  @assertion.assert_eq(queue.peek(), Some(1))?
+  queue.clear()
+  @assertion.assert_eq(queue.peek(), None)?
+}
+
+/// Pops the first value from the queue.
+/// 
+/// # Example
+/// ```
+/// let queue : Queue[Int] = Queue::[1, 2, 3, 4]
+/// let first = queue.pop()
+/// ```
+pub fn pop[T](self : Queue[T]) -> T {
+  match self.first {
```

Comment:
make it consistent with `peek`, we can provide `pop_exn` and `pop` (return Option[T])